CESM-GC updates for cam_development

lizziel opened the pull request:
This is a dummy PR to show documentation of CESM-GC changes to merge into the main CAM development branch. The source branch is a version of CESM-GC_main rebased onto CAM 6.0.034, with all CAM code in stale CESM 2.1.1 branch dropped, and many CESM-GC commits squashed.

The latest cam_development (target) has not been merged into the source branch and there are thus conflicts. This dummy PR also serves to help sort out those conflicts.

fritzt committed:
Feat: Changes to include GEOS-Chem as a chemistry option for CESM

(1) This introduces an option to compile GEOS-Chem src files.

(2) Interface is inexistant so far

(3) CAM still uses the Terminator chem package to satisfy dependencies

Reproducing commit from MSL - Jan 19, 2018

fritzt committed:
Squashed of 50+ commits from Thibaud Fritz

Feat: Update config files to include GEOS-Chem option for CAM
 (1) Update config files in cam/bld/
 (2) Update config files in cam/cime_config
Feat: Add customized chemistry.F90 for GEOS-Chem:
  *Reproducing commit from MSL - Mar 6, 2018
Feat: Initial implementation of compilable code
 (1) The code can now be compiled using a CAM 4.0 + "dummy" GEOS-Chem
     component set (compset FGC). This is based on the "tropospheric
     mozart" set.
  *Reproducing commit from SDE - May 17, 2018
Feat: Make radiatively-active species into chemical constituants
 (1) Added N2O, CH4, CFC11, CFC12
Feat: Get tracers and species from input.geos and KPP
 (1) Tracer lists are now acquired from the input files and from the
     pre-built KPP mechanism
  *Reproducing commit from SDE - May 19, 2018
Feat: Add additional flags when compiling GEOS-Chem
Fix: Update folder name following version 12.6.0
Feat: Allocate major state variables (Input, Met, Chem)
 (1) Allocating GEOS-Chem state variables
 (2) Setting input.geos path as a module variable
  *Reproducing commit from SDE - May 19, 2018
Chore: Cleanup and capitalization
Feat: Initialize core GEOS-Chem modules
 (1) Initializes ChmState, GrdState and MetState objects
 (2) Grid objects are currently initialized identically. However, they
     should be chunk-dependent. The grids are set based on the lat/lon
     edges for now, using GEOS-Chem's SetGridFromCtrEdges subroutine
 (3) Initializes Input_Opt
 (4) Added subroutine to update time-steps within GEOS-Chem
 (5) Added error trapping statements
Feat: Add initialization of Linoz module and passing CHEM_INPUTS
 (1) Initializing Linoz module
 (2) Passing CHEM_INPUTS to Input_Opt
 (3) Remove incorrect overwrite of NTracers
Feat: Initialize and cleanup State_Chm and State_Met
Feat: Initialize Drydep_Mod
 (1) Initialize Drydep_Mod
 (2) Update .exclude file to include GeosCore/drydep_mod.F
Feat: Add Init_Error to chem_init for GEOS-Chem
Feat: Initialize and cleanup GEOS-Chem modules
 (1) Add calls to module initialization and cleanup subroutines
 (2) Update .exclude file
Feat: Get grid area and Ap and Bp values from CAM
 (1) Get grid area from CAM interface
 (2) Pass CAM's Ap and Bp values to GEOS-Chem's Pressure_Mod
 (3) Call to Cleanup_Pressure
Feat: Affect data to State_Chm
Feat: Initialize chemistry, add missing cleanup routines
Feat: Initialize pressure transfer
Feat: Add first call to chemistry
Feat: Register all tracers using species database
 (1) Register species
 (2) Initial step towards the update of State_Met
Feat: Fill key variables in State_Met
Feat: Add PBL height transfer capability
 (1) Enforce lat/lonMidArr to be R4 not FP
Feat: Enforce correct tracer registration
Feat: Add initialization of short-lived species
Feat: Update surface area during chemistry loop
Feat: Use reference MMRs when queried for ICs
 (1) Zero reference MMRs at initialization
 (2) Use reference MMRs when queried for initial conditions
Feat: Groundwork for correctly reporing H2O tendencies
 (1) Add indices to key species
Feat: Add all advected species to output
 (1) Add advected species to output
 (2) Force precision of H2O to be r8
 (3) Fixed capitalization of variables
Feat: Make short-lived species persist between time steps
Feat: Set GEOS-Chem data from CAM
 (1) Date information is now propagated to GEOS-Chem. Algorithm needs
     tweaking (seems to give a date 1 minute earlier than it should).
     Also currently force the year to be 2000 as the default CAM year
     is 0000.
Feat: Disable stratospheric aerosols in GEOS-Chem
Feat: Progate more met fields to State_Met using CAM data
Feat: Change NY to PCOLS for consistency and add step count tracking
Feat: Increased default list of diagnostics for GEOS-Chem
Feat: Implement GEOS-Chem wet deposition + Estimate of 2D cloudiness
 (1) Implemented GEOS-Chem wet deposition
 (2) CAM does not provide an estimate of 2D cloud cover, so we use an
     estimate (maximum 3D cloudiness in vertical column). This matches
     the technique originally used for GCAP in GEOS-Chem.
 (3) Fix Z0 not being assigned
Feat: Move SLS information to chem_mods + Emission module
 (1) Move tracer and SLS information out to chem_mods
 (2) Added temporary NO source to provide useful output
 (3) Initial code stub for an emissions module
Fix: Fix reading of tracers in input.geos
 (1) Tracers in input.geos
 (2) Fix indentation
Chore: Now compiling GEOS-Chem with -DMODEL_
Feat: Added wet/(dry) deposition. See more comments below!
 (01) Added definition of MaxTropLev, MaxStratLev and MaxChemLev. This
      will require some additional thought
 (02) As of right now, Linoz is turned off as the current grid doesn't
      extend past 30 km, causing segmentation when running with Linoz
      turned on
 (03) Added code that's commented out right now for future development
       (HEMCO, Olson landmap, convection, ...)
 (04) Added field descrition and unit of State_Met variables when
      converted from CESM to State_Met
 (05) Updated compution of cloud optical depth. Following what MOZART
      does.
 (06) Now correctly initializing InChemGrid to .True. everywhere
 (07) Added newDay, newMonth variables
 (08) Added wet deposition
 (09) Added dry deposition. However, it is purely a software right now,
      as it runs without initializing land types
 (10) Uniformization of variables
Chore: Remove debug
Feat: Deal with compile-time definition of GEOS-Chem dry dep. species
 (1) Update ChemNamelist to be able to read deposition species from .xml
     files and filtering these with the species list from mo_sim_dat.F90.
     This last file lists all GEOS-Chem species and is only read at
     compile time and is not compiled since GEOS-Chem tracers and species
     are defined at run time from input.geos and KPP files.
 (2) Update geoschem.xml to now list dry and wet deposition species
Feat: Implement dry deposition in CESM/GC
 (1) Different options are available to compute dry deposition
     velocities:
    1. All deposition velocities are computed from GEOS-Chem using data
       from HEMCO
    2. CLM passes dry deposition velocities over land and ocean and ice
       velocities are computed from GEOS-Chem using data from HEMCO
    3. CLM passes dry deposition velocities over land and ocean and ice
       velocities are computed in a similar way as MOZART
Feat: Add missing file from last commit to replicate MOZART's dry dep. calculations
Feat: Add Externals to download GEOS-Chem source code
Feat: Add mo_chem_utils.F90 required for mo_drydep_mod
Feat: Update .exclude
Feat: Import CLM4.0/4.5/5.0 data and use it for GEOS-Chem
 (1) Import data from CLM
 (2) Add compset to run GEOS-Chem with CLM4.5/5.0
 (3) Add routine getLandTypes to convert CLM to OlsonLandMaps. More work
     is needed to convert CLM4.5 land types
 (4) CAM's building procedure has been modified to include the right
     version of CLM
 (5) Modify .exclude
Style: Fix typos that got introduced when converting commits to fork
Feat: Modify Externals_CAM.cfg to download GEOS-Chem code
Revert "Feat: Modify Externals_CAM.cfg to download GEOS-Chem code"
Feat: Update .gitignore to ignore GEOS-Chem repo

jimmielin committed:
Squashed 5 commits from Haipeng Lin

Feat: Implement HEMCO_CESM within CESM-GC (Initial attempt in syncing CAM sources)
Feat: HEMCO_CESM code integration (full). Add mo_sim_dat and mo_tracname
Fix: Call set_sim_dat from chemistry.F90 to pass solsym to HEMCO_CESM.
Fix: Do not assign other variables within mo_sim_dat -- to discuss
Fix: Comment out mo_sim_dat except solsym, expand solsym size to nTracersMax
*This might not be needed after all as HEMCO_CESM now reads tracer names from chem_mods. But fixing this so it doesn't infinite loop.

fritzt committed:
Squashed 2 commits from Thibaud Fritz

Feat: Add Olson / CLM land type mapping for CLM4.5/5.0
Feat: Add FAST_JX_DIR in Input_Opt. Remove hard-wired paths in fast_jx

lizziel committed:
Squashed 4 commits from Lizzie Lundgren

Change GEOS-Chem source to CESM-GC GitHub organization and branch name
Use GEOS-Chem branch feature/13.0.0+CESM
Update HEMCO interfaces directory name in configure
Partial updates to enable GEOS-Chem dev/13.0.0 to build in CESM

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

fritzt committed:
Squashed 60+ commits from Thibaud Fritz

Feat: Increase nchem_adv
Feat: Define GEOS-Chem deposition species in geoschem.xml
Feat: Update solsym based on gckpp_Monitor
 (1) solsym needs to be defined explicitely, as it is parsed at compile
     time.
Feat: Update .exclude to now include input_mod and pops_mod.F90
Feat: Changes required to make GEOS-Chem 13.0 work with CESM
Feat: Set State_Met%Salinity and Iodide to 0. Will be read from HEMCO
Feat: Compute isLand/Water/Ice/Snow after AirQnt + Get fields from HEMCO
 (1) Get IODIDE surface concentration and SALINITY from HEMCO
Feat: Kludge gaseous species names to match CESM restart file:
 (1) Bromine, chlorine, iodine and lumped species are not in CESM restart
     file. We can choose to load these from a GEOS-Chem restart file but
     this is future work.
Feat: Rename CAM's drydep_mod to aer_drydep_mod
Feat: Print dep. lists when getting from namelist
Feat: Allow MAM to compile with CESM-GC
Feat: Remove unused use statement
Feat: Updates to make MAM4 interchange data with CESM-GC
 Compile time changes:
 (1) prescribe_aerosols is set to FALSE when $chem contains geoschem
 (2) Replace aerosol deposition species with MAM4 aerosols in nml files
 (3) Remove pp_geoschem/aero_model. Currently, GEOS-Chem only works
     with MAM(4) turned on. If MAM is to be switched off in the future,
     then the compiler should choose to compile pp_geoschem/aero_model
     (which is just a dummy, almost empty file) over
     modal_aero/aero_model.F90
 Code changes:
 (1) Make xname_massptr a public and protected variable. It is used to
     establish a mapping between MAM4 and GEOS-Chem aerosols
 (2) Add map2MAM4(:,:) in chem_mods and set number of fixed species to
     6 rather than 4
 (3) Add dummy constituents for GEOS-Chem aerosols. Ideally, aerosols
     that are inherited from MAM should not need to be defined as a
     constituent. For instance, BCPI is inherited from bc_a* whereas
     NIT is not included and needs to be defined as a constituent (in
     order to be advected)
 (4) Establish mapping between MAM4 and GEOS-Chem aerosols. Currently,
     POM and SOA are not mapped. More work is needed here to figure out
     which GEOS-Chem aerosols should be used to map those.
 (5) Reverse mapping (aka from GEOS-Chem back to MAM) is not established
     yet.
 (6) Clip QI and QL to 1.0E-05 as low values (I have found values as low
     as 1.0E-141) can make GEOS-Chem blow up in sulfate_mod.
 (7) Add empty m_spc_id.F90, required by MAM4
 (8) In mo_chem_utils, replace tracerNames by tracname. This is required
     as MAM aerosols are not in tracerNames (which only includes
     GEOS-Chem species). Add get_inv_ndx. I believe that this change
     temporarily breaks dry deposition as performed by MOZART. More
     investigation is needed here.
 (9) Add subroutine in mo_drydep to load in land types which is required
     for MAM4 to perform its own dry deposition
 (10) Add MAM4 aerosols to solsym and define inv_lst (list of fixed
      species)
 ----
 TODOs:
 (1) Establish reverse mappings (GEOS-Chem -> MAM4)
 (2) Check MOZART's dry deposition routines.
 (3) As of right now, MAM4 performs dry deposition. Would it be possible
     to let GEOS-Chem perform dry deposition of aerosols?
 (4) What to do regarding pure GEOS-Chem aerosols (e.g. NIT)?
Chore: Fix indentation + capitalization
Feat: Add option to get land types from HEMCO
Fix: Aerosol mapping requires more indices in MMR_Beg.
Remove custom emissions
Feat: Add emissions to CESM-GC
Fix: Prevent emissions from adding up in CESM-GC. Reset cam_in%cflx
Feat: Modify aero_model.F90 to let GEOS-Chem handle its own aerosols
Fix: Resolve circular dependency
Feat: Rename HCO_surf_salinity to HCO_salinity
 (1) The physics buffer does not allow to store strings as long as
     `HCO_surf_salinity` and would, for instance, store
     `HCO_surf_salinit`. To avoid confusion, we rename it to
     `HCO_salinity`, same for `HCO_iodide`.
Feat: Allow CESM-GC to output dry deposition velocity used by GEOS-Chem
 (1) Add addfld and outfld statements for each dry deposition species,
     using State_Chm%DryDepVel
 (2) Fix to let chemistry.F90 when LANDTYPE_HEMCO is 1.
Feat: Uncomment call to DO_MIXING to allow for dry deposition
Feat: Change IC condition file
 (1) Renamed some species into their GEOS-Chem equivalent. This new IC
     file is a mere copy of the default file, with some species renamed
Fix: Replace species number with `gas_pcnst`
Feat: Reindexed GEOS-Chem species: tracers, MAM aerosols, SL species
 (1) This reindexing is necessary as MAM requires that there is a linear
     mapping (constant offset) between solsym and constituents.
 (2) adv_mass is now properly defined in mo_sim_dat, using GEOS-Chem
     values for all species, except MAM aerosols.
Feat: Add runtime option to using GEOS-Chem wet deposition
Feat: Add files to allow for NEU wet deposition scheme
 (1) Add mo_neu_wetdep (copied from MOZART)
 (2) Add additional files (mo_mean_mass, mo_setinv, gas_wetdep_opts)
 (3) Add new field in mo_neu_wetdep (WDRATE), corresponding to the wet
     removal rate in kg/s
Feat: Add NEU WD routines + MAM gas-aerosol exchange
 ( 1) Added NEU wet deposition routines. Option to choose between NEU and
      GEOS-Chem wet deposition scheme is available at runtime. See commit
      #a8d642c
 ( 2) Added diagnostic fields for wet deposition. These match the names
      of the same quantities in MOZART
 ( 3) Added MAM gas-aerosol exchange
 ( 4) Rename field DDVel and SurfFlux into DepVel and DepFlux. Might need
      to change those in the future to match those in MOZART
 ( 5) Established MAM4 aerosol to GEOS-Chem species mapping based on
      discussion with Louisa Emmons
 ( 6) Added inverse mapping map2GCinv hat maps tracers onto constituent.
      This is now used in chem_emissions.
 ( 7) Added map2chm which maps solsym onto GEOS-Chem species and
      constituents
 ( 8) Remove special handlings when defining constituents. Variable names
      are now changed in the IC file.
 ( 9) Compute CSZAmid field
 (10) Fix bug where qH2O was flipped vertically
Feat: Get UVALBEDO from HEMCO
 *CAM only defines albedo during daytime. However, UVALBEDO is only used
 for photolysis, which should be fine.
Feat: Define MaxTrop/StratLev from grid + Add Set_H2O_Trac
Feat: Add OH and JRates diagnostics.
Feat: Get H2O mmr from specific humidity
 (1) GEOS-Chem's water mass mixing ratio is derived from CAM's specific
     humidity
 (2) Additionally, the JRates diagnostics are only computed if they are
     an output field
Chore: Remove unnecessary comment
Fix: Unit check failed because 'kg/ kg dry' /= 'kg/kg dry'
Feat: Remove special handlings as we now use GEOS-Chem IC
 (1) Previously GEOS-Chem species were modified to account for the
     species in the CAM restart file (e.g. HNO2 vs HONO). Now, we just
     use GEOS-Chem initial conditions that have been regridded to
     ~0.9x1.25
Feat: Add option to pass H2O tendency to Q + change some addfld
Feat: Set State_Met%OMEGA equal to state%omega
Feat: Allow CESM-GC to compute overhead ozone from State_Chm%Species
Feat: Allow CESM-GC runs to be restartable
Feat: Implement 3-D emissions.
 (1) Surface emissions (layer PVER) are passed to cam_in%cflx while the
     non-surface emissions are added directly to the chemical tendencies
 (2) Dry deposition fluxes are now passed to cam_in%cflx such that:
     cam_in%cflx = eflx - dflx
 (3) Remove call to DO_MIXING as this is now handled by CAM.
Feat: Replace tracerNames with solsym in chem_implements_cnst
Feat: Changes required by GEOS-Chem 13.0.0
Feat: Move compile time flags to run time flags for DD velocities
Feat: Read OMOC from HEMCO and store in State_Chm
Feat: Move all GEOS-Chem related diagnostics to cesmgc_diag_mod.F90
Feat: Add missing initialization of ThisLoc and ErrMsg
Feat: Add option to output State_Met in cesmgc_diag_mod.F90
Feat: Update GEOS-Chem branch in Externals_CAM.cfg to be `CESM`
Feat: Replace pp_geoschem with geoschem
Feat: Reset MAM-inherited aerosols to 0 before mapping
 (1) Resetting MAM-inherited aerosols in State_Chm%Species allows these
     aerosols to not accumulate over time. Otherwise, we would have
     for instance:
         BCPI = BCPI + MAM4,
     which would accumulate overtime if BCPI was not reset
 (2) Lowered the threshold on QI and QL to 1.0E-10 rather than 1.0E-05
Feat: Add MAM aerosols to cesmgc_diag_mod
Feat: Add lightning emissions
Feat: Add MEGAN emissions
Feat: Symbolic link to mozart/mo_lightning
Feat: Add symbolic link to MOZART files, rather than copy
Feat: Add lightning, MEGAN, aerosol and fire emissions
Feat: Remote WDRATE_* from CAM history fields
Fix: Molar weight pFe
Fix: Add to_upper to tracerNames, required since 'pFe' /= 'PFE'
Chore: Try lower and uppercase tracernames when getting data from pbuf
Fix: Handle cases where ncol /= pcols
Chore: Add note explaining why DQRLSAN needs to be zero in top layer
Feat: Pass GEOS-Chem aerosol emissions to MAM tendencies
Fix: Fix typo SO4S vs SOAS
Feat: Enforce lq('Q') to be true if passing H2O tendencies:
 (1) lq(cQ) = .True. if applyQTend is True
 (2) iFirstCnst is now a variable in chem_mods
 (3) Reset cam_in%cflx for all species (including MAM aerosols)
Feat: Only diagnose wet deposition rates of soluble species
Feat: Skip emissions on first time step before HEMCO fills in data
Feat: Add chem_is-like function in mo_chm_utls
Fix: Fix typo
Feat: Add or rename some output fields:
 (1) Add option to save out chemical tendencies (kg/kg/s or kg/s)
 (2) Rename output fluxes to match CAM-Chem
Feat: Add _CLFX and other CAM-Chem like diagnostics to CESM-GC
Feat: Add new GC compsets that mimic CAM-Chem compsets:
 (1) New compsets
 (2) Enforce NTHRDS_ATM to be 1 when running with GEOS-Chem
Feat: Add new use_cases XML entries for newly added GC compsets
Feat: Turn on rad_prog_ozone for GEOS-Chem chemistry
Feat: Implement CAM-Chem like diagnostics for CESM-GC:
 (1) Rename diagnostics to match CAM-Chem's
 (2) Only perform diagnostics computations if field is an output.
     Use hist_fld_active whenever a call to outfld is performed.
Feat: Make sure that GC's SOAGX species is not picked up by MAM
Fix: Remove debug statement in cesmgc_diag_mod
Feat: Add mmr tendencies for MAM aerosols from chemistry
Feat: Replace DO loops with array element-wise operations
Feat: Set NTHRDS_*=1 when using a GEOS-Chem compset
Feat: Make SDYN compset with GEOS-Chem have 56 levs
Feat: Pass chemical tendencies to MAM aerosols:
 (1) Changes in aerosol concentrations due to GEOS-Chem processing is
     now passed to chemical tendencies for MAM aerosols
 (2) Wet deposition of MAM-inherited aerosols is not performed in
     GEOS-Chem. To do so, we define a new logical `WD_ExternalDep` in
     SpcInfo
Feat: Remove mass-weighted OH concentration diagnostics
Feat: Add CO2 as constituent
Feat: Compute CO2 chemical tendencies:
 (1) GEOS-Chem overwrites CO2 concentrations at every time step. We thus
     set State_Chm%Species(:,:,:,iCO2) = 0.0e+00 before chemistry and
     compute chemical tendencies based on how much CO2 has been produced.
 (2) Cleanup
Feat: Apply CAM-Chem surface boundary conditions for long-lived species
Feat: Remove hardwired path to GEOS-Chem CHEM_INPUTS directory
Feat: Remove hard-wired paths for GEOS-Chem inputs:
 (1) All input files are now (one-time) copied from geoschem_src/run/CESM
     to Buildconf/camconf. Then to the run folder every time builnml is
     called
Feat: Read from input.geos
Feat: Read input.geos rather than hard-wiring input options
Feat: Add check to make sure that solsym is following list of GC tracer
Chore: Clean up in short_lived_species
Fix: Fix wrong reference MMR
Chore: Cleanup, remove unused MWRatio and longNames
Feat: Update .exclude to exclude gosat and tccon from GEOS-Chem folder
Feat: Get strat_chem data from HEMCO
Feat: Add timers around DO_CHEMISTRY + Fix diagnostic in WetDep
Feat: Remove PSO4 diagnostic (write statement)
Fix: Fix diagnostics of emissions
Feat: Remove option to let GEOS-Chem perform wet scavenging.
 (1) The only option to perform convective scavenging in CESM-GC is
     currently to rely on the NEU scheme, which handles large-scale
     precipitation and convective scavening. Since the variables
     required by the GEOS-Chem convective routines cannot be extracted
     easily, we rely on the NEU scheme to perform washout due to both
     LS precipitation and convective scavenging for gases AND aerosols.
 (2) A next commit will follow to specify Henry coefficients of aerosols.
     The best option is to probably set something similar as HNO3
Feat: Make non-MAM GC aerosols "gas-like" species for wet&dry dep
 (1) Move all GC aerosols from aer_drydep_list to drydep_list and from
     aer_wetdep_list to gas_wetdep_list (see previous commits too)
 (2) This allows to remove the 'GC_AER_' terminology for GC aerosols
 (3) Aerosol emission fluxes (e.g. BCPI) are now passed directly to
     MAM (e.g. flux to bc_a4). This allows for the removal of some
     interfacing code in cesmgc_emissions_mod.F90
 (4) Fix diagnostic naming issues (e.g. replaced WD_HBR with WD_HBr)
 (5) Remove extra bit of code in cesmgc_diag_mod dealing with wet dep
     in GEOS-Chem, which had been removed in previous commits
 (6) In mo_neu_wetdep.F90, make all GC aerosols be removed like HNO3
 (7) Revert previous changes in modal_aero that used to deal with
     GC aerosols in aer_drydep_list

Signed-off-by: Thibaud Fritz <fritzt@mit.edu>

jimmielin committed:
Feat: Write ZPJ values for RXN_NO2, RXN_O3_1/2a into HCO_IN_JNO2, HCO_IN_JOH for HEMCO ParaNOx extension.

Fix: fldname_ns fix to FIELDNAME

fritzt committed:
Squashed 4 commits from Thibaud Fritz

Feat: Uniformize calls to HEMCO
Fix: Replace SO4S with SO4s
Fix: Fix XML file name
Feat: Capitalize all GC constituents name, and other updates:
(1) This is required because boundary conditions are stored with
    capitalized name. I would rather follow the same terminology
    as MOZART (with capitalized constituents) rather than modify
    the interface with flbc.
(2) Update geoschem.xml to match FC2000climo compset
(3) Fix bug where lght_no_prod_factor was not applied in CESM-GC
(4) Set ext_frc_specifier to '' for CESM-GC compsets as we rely on
    HEMCO

jimmielin committed:
Feat: Now retrieve fields from HEMCO using 2-D pbuf fields

fritzt committed:
Squashed 10 commits from Thibaud Fritz

Fix: Pass surface J-Rates to HEMCO for PARANOX
Fix: Update MMR_Beg for MAM aerosols:
(1) This fix avoids a slow, but steady build up of some MAM aerosols
Fix: Make sure GEOS-Chem (aerosol phase) MSA is not picked up by MAM
Feat: Diagnose constituents as mol/mol apart from MAM aerosols
Feat: Update namelist_defaults s.t. geoschem mimics trop_strat_mam4vbs
Feat: Update geoschem.xml to mimic history_* XML variables
Feat: Add history_* options to save out fields in cesmgc_diag_mod
Feat: Don't make solsym fixed size, but rather of size gas_pcnst
Feat: Change nadv_chem from 240 to 238
Feat: Add more history_* options to save out fields (mimic CAM-Chem)

Signed-off-by: Thibaud Fritz <fritzt@mit.edu>

fritzt committed:
Squashed 18 commits from Thibaud Fritz

Feat: Revert mo_drydep to MOZART's version + Update chemistry.F90
 (1) Note that most of the mo_drydep routines are unused. We only need
     to load in landtypes for MAM
Feat: Add aerosol number emissions in CESM-GC
 (1) Emission diameters are taken from:
     Emmons, Louisa K., et al. "The chemistry mechanism in the
     Community Earth System Model version 2 (CESM2)."
     Journal of Advances in Modeling Earth Systems 12.4 (2020).
Feat: Lump GEOS-Chem SO4 with so4_a* and H2SO4
Fix: Remove hard-wired number emissions as this has been moved to HEMCO
Feat: Update .xml files with GEOS-Chem chemistry (MEGAN)
Feat: Add MEGAN emissions with GEOS-Chem species
Fix: Update IF condition for SRF diagnostics
Feat: Add hemco_nl in build-namelist
Feat: Rename geoschem.xml to 2000_geoscheml.xml (FC2000climo_GC)
Feat: Add 2010_geoscheml.xml (FC2010climo_GC)
Feat: Update config_component to reflect recent changes to xml files
Feat: Add SOA mapping between MAM and GEOS-Chem complex SOA option
 (1) Add mapping
 (2) Extend number of constituents
Feat: Add AEIC organic species to external forcings list
Feat: Add GEOS-Chem surface boundary conditions
Feat: Update .gitignore to exclude GEOS-Chem source code
Fix: Fix for sulfur mapping between CESM2 and GEOS-Chem
 (1) SO4_gasRatio and binRatio are in mol/mol, not kg/kg
Feat: Update .exclude for GEOS-Chem 13.1
Fix: Update in-cloud opticaldepth with cloud fractions

jimmielin committed:
Feat: Retrieve ParaNOx deposition fluxes from HEMCO

lizziel committed:
Fix: correct improper merge conflict handling introduced during rebase

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

fritzt committed:
Feat: Return in sox_cldaero_mod if running with GEOS-Chem

This prevents the code from double counting in-cloud sulfur oxidation

Signed-off-by: Thibaud Fritz <fritzt@mit.edu>

lizziel committed:
Merge branch 'cam_development' into CESM-GC+cam_development

NOTE: Update to add AEIC organic species to external forcings list was
      removed in this merge. It is required for using HEMCO with CAM-Chem
      since all 3D emissions fluxes applied in CAM-Chem must be
      hard-coded. The update is not required for CESM-GC and so adding it
      will be done later.

# Conflicts:
#	.gitignore
#	bld/build-namelist
#	bld/config_files/definition.xml
#	bld/configure
#	bld/namelist_files/namelist_defaults_cam.xml
#	bld/perl5lib/Build/ChemNamelist.pm
#	cime_config/buildcpp
#	cime_config/config_component.xml
#	cime_config/config_pes.xml
#	src/chemistry/modal_aero/aero_model.F90
#	src/chemistry/modal_aero/modal_aero_gasaerexch.F90
#	src/chemistry/pp_trop_strat_mam4_vbs/chem_mods.F90
#	src/chemistry/pp_trop_strat_mam4_vbs/mo_sim_dat.F90
#	src/chemistry/pp_trop_strat_mam4_vbsext/chem_mods.F90
#	src/chemistry/pp_trop_strat_mam4_vbsext/mo_sim_dat.F90
#	src/control/cam_comp.F90
#	src/control/camsrfexch.F90
#	src/cpl/atm_import_export.F90
#	src/cpl/mct/cam_cpl_indices.F90
#	src/physics/cam/phys_control.F90

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel commented:
@fritzt and @jimmielin, this dummy PR is a precursor to making a PR on the upstream CAM and discussing changes with NCAR. Could you look over the file changes and tell me if anything should be excluded from the PR, or if anything looks wrong?

A bit of information on what you are looking at:
The source branch is a modified version of CESM-GC_main. I squashed 223 commits from the three of us into 12 commits total, and then dropped CESM commits between CAM 6.0.034 and the CESM 2.1.1 version Thibaud started with. I then merged the latest cam_development as of yesterday into it and did conflict resolution. The only features I have removed/disabled so far from the original changes are (1) change HEMCO from on to off by default (thanks for the very clear message about that in the code!), and (2) remove AEIC species from the external forcing list.

I expect that I should change more before opening the ESCOMP PR and I'm going to look at the files now. Let's discuss on this thread.

fritzt commented on src/chemistry/pp_trop_strat_mam4_vbs/mo_sim_dat.F90:
```diff
@@ -246,14 +246,14 @@ subroutine set_sim_dat
                           1739,1845,1876,1900,1935,1993,2054,2080 /)
 
-      extfrc_lst(: 16) = (/ 'so4_a1          ','bc_a4           ','SVOC            ','bc_a1           ','CO              ', &
+      extfrc_lst(: 24) = (/ 'so4_a1          ','bc_a4           ','SVOC            ','bc_a1           ','CO              ', &
```
The indexing is wrong here. There are only 16 elements. Similar indexing errors also appear below and in other `mo_sim_dat.F90`

lizziel replied:
Thanks for catching that. I manually reverted the changes and forgot that index.

fritzt commented on src/chemistry/geoschem/gas_wetdep_opts.F90:
```diff
@@ -0,0 +1,79 @@
+    if (( gas_wetdep_cnt>0 ).and. &
+       ( .not.(gas_wetdep_method=='MOZ' .or. &
+               gas_wetdep_method=='NEU' .or. &
+               gas_wetdep_method=='GEOS-CHEM' .or. &
```
Just realized this now, but this line "`gas_wetdep_method=='GEOS-CHEM' .or. &`" can be removed. We used to have the option to perform scavenging using the GEOS-Chem approach, but this doesn't work out in CESM2-GC.

lizziel replied:
I'll remove it.

jimmielin commented:
Hi Lizzie,

Thanks so much for working on this. I'm mostly familiar with the HEMCO changes. This all looks good to me. Disabling HEMCO by default is good, for it to work with CAM-chem we still have to work with the CAM-chem folks to add the changes to their emissions code (the changes in https://github.com/jimmielin/HEMCO_CESM/tree/development/SourceMods). We can sort that out in the future.

Best,
Haipeng

fritzt commented:
Hi Lizzie! Thank you for putting all of this together! This looks good to me. Some of the files were copied over from the `mozart` folder and are mostly commented out (e.g. `charge_neutrality.F90`, `chem_prod_loss_diags.F90`, `clybry_fam.F90`, `mo_gas_phase_chemdr.F90`, ...). I'm wondering if it would be worth it to remove all of this commented code. However, we can't just delete them, as some of these files have dependencies with other CAM modules.

lizziel committed:
Fix: correct wrong indexes introduced by manual revert; rm GC wetscav

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel commented:
See new PR at https://github.com/ESCOMP/CAM/pull/484

lizziel committed:
Merge branch 'cam_development' into CESM-GC_rebased_on_cam6_0_034

lizziel committed:
Merge branch 'cam_development' into CESM-GC_rebased_on_cam6_0_034

lizziel committed:
Merge branch 'cam_development' into CESM-GC_rebased_on_cam6_0_034

lizziel committed:
Enable compilation in CESM 2.3 for non-GEOS-Chem case

This commit includes:
 1. Syntax bug fixes in perl file ChemNamelist.pm
 2. Setting perl smartmatch to experimental in ChemNamelist.pm to avoid
    perl version dependent build error for use of '~~' (checks if item
    is in array)
 3. Remove CAM imports lwtgcell, pwtgcell, and lai that are used for
    GEOS-Chem dry deposition velocity calculation. In a future update
    GEOS-Chem will no longer calculate dry deposition velocity when run
    in CESM so these imports are not needed.

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Build fixes for CESM-GC within CESM: non-GEOS-Chem cases

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Update GEOS-Chem interface code for compatibility with CESM 2.3

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>
